File: src/06/src/assembler/parser.ts
import * as fs from 'fs'
import * as path from 'path'

import { A_COMMAND, C_COMMAND, L_COMMAND } from './constants'

class Parser {

  instructions: string[]
  lineCounter: number
  currentCommand: string

  /**
   * 入力ファイルを展開しパースする準備を行う
   */
  constructor(filePath: string) {
    const fileContent = fs.readFileSync(path.resolve(__dirname, filePath), { encoding: 'utf-8' })
    const lines = fileContent.replace(/ /g, '').split(/\n/)
    this.instructions = lines.filter((line) => {
      return line !== '' && line.indexOf('//') !== 0
    });
    this.lineCounter = 0
    this.currentCommand = this.instructions[this.lineCounter]
  }

  /**
   * 入力にまだコマンドが存在するかどうかを判定する
   */
  hasMoreCommands(): boolean {
    return this.instructions.length > this.lineCounter
  }

  /**
   * 次のコマンドを読み込んで現在のコマンドに変更する
   */
  advance(): void {
    if (!this.hasMoreCommands) return
    this.lineCounter = this.lineCounter + 1
    const command = this.instructions[this.lineCounter]
    if (!command) return
    this.currentCommand = command.split('//')[0]
    return
  }

  /**
   * 現コマンドの種類を返す
   */
  commandType(): string {
    if (this.currentCommand.indexOf('@') === 0) {
      return A_COMMAND
    } else if (this.currentCommand.indexOf('(') === 0) {
      return L_COMMAND
    } else {
      return C_COMMAND
    }
  }
  
  /**
   * 現コマンドのシンボルを返す
   */
  symbol(): string {
    if (this.commandType() === C_COMMAND) throw new Error('commandType should be C_COMMAND when call symbol')

    let returnValue = ''
    if (this.commandType() === A_COMMAND) {
      returnValue = this.currentCommand.slice(1)
    }
    if (this.commandType() === L_COMMAND) {
      returnValue = this.currentCommand.slice(1, -1)
    }
    return returnValue
  }

  /**
   * destニーモニックを返却する
   */
  dest(): string|null {
    if (this.commandType() !== C_COMMAND) {
      throw new Error('commandType should be C_COMMAND when call dest');
    }
    if (this.currentCommand.indexOf(';') !== -1) {
      return null;
    }
    return this.currentCommand.split('=')[0];
  }

  /**
   * compニーモニックを返却する
   */
  comp(): string {
    if (this.commandType() !== C_COMMAND) {
      throw new Error('commandType should be C_COMMAND when call comp');
    }
    if (this.currentCommand.indexOf(';') !== -1) {
      return this.currentCommand.split(';')[0];
    }
    return this.currentCommand.split('=')[1];
  }

  /**
   * jumpニーモニックを返却する
   */
  jump(): string {
    if (this.commandType() !== C_COMMAND) {
      throw new Error('commandType should be C_COMMAND when call jump');
    }
    return this.currentCommand.split(';')[1];
  }
}

export default Parser

File: src/Archive/assembler/parser.rs
use std::fs::File;
use std::io::{BufReader, BufRead};
use std::convert::TryInto;

#[derive(PartialEq)]
enum CommandType {
    A_COMMAND,
    C_COMMAND,
    L_COMMAND
}

pub struct Parser {
    instructions: Vec<String>,
    line_counter: usize,
    current_command: String
}

impl Parser {
    pub fn new(f: File) -> Parser {
        let reader = BufReader::new(f);
        let lines: Vec<String> = reader
            .lines()
            .map(|l| l.unwrap())
            .map(|l| match l.find("//") {
                Some(v) => l[0..v].to_string(),
                None => l,
            })
            .filter(|l| !l.is_empty())
            .map(|l| l.trim().to_owned())
            .collect();
        let first_cmd = lines[0].to_string();

        Parser {
            instructions: lines,
            line_counter: 0,
            current_command: first_cmd
        }
    }

    pub fn has_more_commands(&mut self) -> bool {
        self.instructions.len() > self.line_counter
    }

    pub fn advance(&mut self) {
        if !self.has_more_commands() {
            return
        }
        self.line_counter += self.line_counter;

        let command = &self.current_command;
        if command.is_empty() {
            return
        }
    }

    pub fn command_type(&mut self) -> CommandType {
        match &self.current_command[0..1] {
            "@" => CommandType::A_COMMAND,
            "(" => CommandType::L_COMMAND,
            _   => CommandType::C_COMMAND
        }
    }

    pub fn symbol(&mut self) -> String {
        let str = match self.command_type() {
            CommandType::A_COMMAND => &self.current_command[1..],
            CommandType::L_COMMAND => {
                let start: usize = self.current_command.find("(").unwrap();
                let end: usize = self.current_command.find(")").unwrap();
                &self.current_command[start..end]
            }
            CommandType::C_COMMAND => ""
        };
        str.to_string()
    }

    pub fn dest(&mut self) -> &str {
        if self.command_type() != CommandType::C_COMMAND { panic!("abababababababababa"); }

        let is_have_char = match self.current_command.find(";") {
            Some(size) => true,
            None => false
        };

        if is_have_char {
            let tmp: Vec<&str> = self.current_command.split("=").collect();
            tmp[0]
        } else {
            ""
        }
    }

    pub fn comp(&mut self) {
        if self.command_type() != CommandType::C_COMMAND {
            panic!("abababababababababa");
        }
        // TODO implement
    }

    pub fn jump(&mut self) {
        if self.command_type() != CommandType::C_COMMAND {
            panic!("abababababababababa");
        }
        // TODO implement
    }
}
